internal/query/failed_events.go

```diff
@@ -0,0 +1,158 @@
+package query
+
+import (
+	"context"
+	"database/sql"
+	errs "errors"
+	"fmt"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/caos/zitadel/internal/errors"
+	"github.com/caos/zitadel/internal/query/projection"
+)
+
+var (
+	failedEventsTable = table{
+		name: projection.FailedEventsTable,
+	}
+	FailedEventsColumnProjectionName = Column{
+		name:  projection.FailedEventsColumnProjectionName,
+		table: failedEventsTable,
+	}
+	FailedEventsColumnFailedSequence = Column{
+		name:  projection.FailedEventsColumnFailedSequence,
+		table: failedEventsTable,
+	}
+	FailedEventsColumnFailureCount = Column{
+		name:  projection.FailedEventsColumnFailureCount,
+		table: failedEventsTable,
+	}
+	FailedEventsColumnError = Column{
+		name:  projection.FailedEventsColumnError,
+		table: failedEventsTable,
+	}
+)
+
+type FailedEvents struct {
+	SearchResponse
+	FailedEvents []*FailedEvent
+}
+
+type FailedEvent struct {
+	ProjectionName string
+	FailedSequence uint64
+	FailureCount   uint64
+	Error          string
+}
+
+type FailedEventSearchQueries struct {
+	SearchRequest
+	Queries []SearchQuery
+}
+
+func (q *Queries) SearchFailedEvents(ctx context.Context, queries *FailedEventSearchQueries) (failedEvents *FailedEvents, err error) {
+	query, scan := prepareFailedEventsQuery()
+	stmt, args, err := queries.toQuery(query).ToSql()
+	if err != nil {
+		return nil, errors.ThrowInvalidArgument(err, "QUERY-n8rjJ", "Errors.Query.InvalidRequest")
+	}
+
+	rows, err := q.client.QueryContext(ctx, stmt, args...)
+	if err != nil {
+		return nil, errors.ThrowInternal(err, "QUERY-3j99J", "Errors.Internal")
+	}
+	return scan(rows)
+}
+
+func (q *Queries) RemoveFailedEvent(ctx context.Context, projectionName string, sequence uint64) (err error) {
+	_, err = q.client.Exec(fmt.Sprintf("DELETE FROM %s WHERE %s = $1 and %s = $2", projection.FailedEventsTable, projection.FailedEventsColumnProjectionName, projection.FailedEventsColumnFailedSequence), projectionName, sequence)
```

Comment:
same as with the current_sequence exec... %s and $1

Reply:
I prefer the current solution. It avoids sql injection by default